rust_src/src/editfns.rs

```diff
@@ -799,4 +800,58 @@ pub fn char_equal(c1: LispObject, c2: LispObject) -> bool {
     unsafe { downcase(c1 as c_int) == downcase(c2 as c_int) }
 }
 
+/// Insert before point a substring of the contents of BUFFER.
+/// BUFFER may be a buffer or a buffer name.
+/// Arguments START and END are character positions specifying the substring.
+/// They default to the values of (point-min) and (point-max) in BUFFER.
+///
+/// Point and before-insertion markers move forward to end up after the
+/// inserted text.
+/// Any other markers at the point of insertion remain before the text.
+///
+/// If the current buffer is multibyte and BUFFER is unibyte, or vice
+/// versa, strings are converted from unibyte to multibyte or vice versa
+/// using `string-make-multibyte' or `string-make-unibyte', which see.
+#[lisp_fn(min = "1")]
+pub fn insert_buffer_substring(buffer: LispObject, beg: LispObject, end: LispObject) -> LispObject {
+    let buf = get_buffer(buffer);
+    if buf.is_nil() {
+        nsberror(buffer);
+    }
+    let mut buf_ref = match buf.as_live_buffer() {
+        Some(b) => b,
+        None => error!("Selecting deleted buffer"),
+    };
+
+    let get_pos = |pos: LispObject, default: isize| {
+        if pos.is_nil() {
+            default
+        } else {
+            pos.as_number_coerce_marker_or_error();
+            pos.as_fixnum_or_error() as isize
+        }
+    };
+    let mut b = get_pos(beg, buf_ref.begv);
+    let mut e = get_pos(end, buf_ref.zv);
+
+    if b > e {
+        let temp = b;
+        b = e;
+        e = temp;
+    }
+
+    if !(buf_ref.begv <= b && e <= buf_ref.zv) {
+        args_out_of_range!(beg, end);
+    }
+
+    let mut cur_buf = ThreadState::current_buffer();
+    unsafe {
+        set_buffer_internal_1(buf_ref.as_mut());
+        update_buffer_properties(b, e);
+        set_buffer_internal_1(cur_buf.as_mut());
+        insert_from_buffer(buf_ref.as_mut(), b, e - b, false)
+    };
+    Qnil
```

Comment:
Oh, btw. You don't have to return nil, you can just not return anything at all.

Reply:
Hmmm. I'm not sure we should do this.

Reply:
I will merge this. If we decide we return nothing instead of `Qnil`, I will change this for all functions. But I want this to be consistent.